jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DoSHandler.java

```diff
@@ -0,0 +1,792 @@
+     */
+    public DoSHandler(@Name("maxRequestsPerSecond") int maxRequestsPerSecond)
+    {
+        this(null, maxRequestsPerSecond, -1);
+    }
+
+    /**
+     * @param trackerFactory Factory to create a Tracker
+     */
+    public DoSHandler(@Name("trackerFactory") Tracker.Factory trackerFactory)
+    {
+        this(null, trackerFactory, null, -1);
+    }
+
+    /**
+     * @param getId Function to extract an remote ID from a request.
+     * @param maxRequestsPerSecond The maximum requests per second allows per ID.
+     * @param maxTrackers The maximum number of remote clients to track or -1 for a default value. If this limit is exceeded, then requests from additional remote clients are rejected.
+     */
+    public DoSHandler(
+        @Name("getId") Function<Request, String> getId,
+        @Name("maxRequestsPerSecond") int maxRequestsPerSecond,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        this(null, getId, new FillingBucketTrackerFactory(maxRequestsPerSecond), null, maxTrackers);
+    }
+
+    /**
+     * @param getId Function to extract an remote ID from a request.
+     * @param trackerFactory Factory to create a Tracker
+     * @param rejectHandler A {@link Handler} used to reject excess requests, or {@code null} for a default.
+     * @param maxTrackers The maximum number of remote clients to track or -1 for a default value. If this limit is exceeded, then requests from additional remote clients are rejected.
+     */
+    public DoSHandler(
+        @Name("getId") Function<Request, String> getId,
+        @Name("trackerFactory") Tracker.Factory trackerFactory,
+        @Name("rejectHandler") Request.Handler rejectHandler,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        this(null, getId, trackerFactory, rejectHandler, maxTrackers);
+    }
+
+    /**
+     * @param handler Then next {@link Handler} or {@code null}
+     * @param getId Function to extract an remote ID from a request.
+     * @param trackerFactory Factory to create a Tracker
+     * @param rejectHandler A {@link Handler} used to reject excess requests, or {@code null} for a default.
+     * @param maxTrackers The maximum number of remote clients to track or -1 for a default value. If this limit is exceeded, then requests from additional remote clients are rejected.
+     */
+    public DoSHandler(
+        @Name("handler") Handler handler,
+        @Name("getId") Function<Request, String> getId,
+        @Name("trackerFactory") Tracker.Factory trackerFactory,
+        @Name("rejectHandler") Request.Handler rejectHandler,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        super(handler);
+        installBean(_trackers);
+        _getId = Objects.requireNonNullElse(getId, ID_FROM_REMOTE_ADDRESS);
+        installBean(_getId);
+        _trackerFactory = Objects.requireNonNull(trackerFactory);
+        installBean(_trackerFactory);
+        _maxTrackers = maxTrackers < 0 ? 10_000 : maxTrackers;
```

Comment:
Could we just default to unlimited here? 10k seems a bit arbitrary.

And if you exceeded this limit you've DOSed yourself because you cannot handle any new requests.

Reply:
If the default is unlimited, then we will just trigger the standard security warnings about being unbounded.   
If we exceed this limit, we are not immediately DOS'd as we start early evicting idle trackers.